hp_list/main.c

```diff
@@ -14,6 +14,129 @@
 #include <threads.h>
 
+static atomic_uint_fast64_t deletes = 0, inserts = 0;
+
+#ifdef RUNTIME_STAT
+
+/*
+ * Reference :
+ * A more Pragmatic Implementation of the Lock-free, Ordered, Linked List
+ * https://arxiv.org/abs/2010.15755
+ */
+
+enum {
+    TRACE_nothing = 0,
+    TRACE_retry,     /* the number of retries in the __list_find function. */
+    TRACE_contains,  /* the number of wait-free contains in the __list_find
+                        function  that curr pointer pointed. */
+    TRACE_traversal, /* the number of list element traversal in the __list_find
+                        function. */
+    TRACE_fail,      /* the number of CAS() failures. */
+    TRACE_del, /* the number of list_delete operation failed and restart again.
+                */
+    TRACE_ins, /* the number of list_insert operation failed and restart again.
+                */
+    TRACE_inserts, /* the number of atomic_load operation in list_delete,
+                      list_insert and __list_find. */
+    TRACE_deletes  /* the number of atomic_store operation in list_delete,
+                      list_insert and __list_find. */
+};
+
+struct runtime_stat {
+    atomic_uint_fast64_t retry, contains, traversal, fail;
+    atomic_uint_fast64_t del, ins;
+    atomic_uint_fast64_t load, store;
+};
+static struct runtime_stat stats = {0};
+
+#define CAS(obj, expected, desired)                                          \
+    ({                                                                       \
+        bool __ret = atomic_compare_exchange_strong(obj, expected, desired); \
+        if (!__ret)                                                          \
+            atomic_fetch_add(&stats.fail, 1);                                \
+        __ret;                                                               \
+    })
+#define ATOMIC_LOAD(obj)                  \
+    ({                                    \
+        atomic_fetch_add(&stats.load, 1); \
+        atomic_load(obj);                 \
+    })
+#define ATOMIC_STORE_EXPLICIT(obj, desired, order)  \
+    do {                                            \
+        atomic_fetch_add(&stats.store, 1);          \
+        atomic_store_explicit(obj, desired, order); \
+    } while (0)
+#define TRACE(ops)                           \
+    do {                                     \
+        if (TRACE_##ops)                     \
+            atomic_fetch_add(&stats.ops, 1); \
+    } while (0)
+
+static void do_analysis(void)
+{
+    __atomic_thread_fence(__ATOMIC_SEQ_CST);
+#define TRACE_PRINT(ops) printf("%-10s: %ld\n", #ops, stats.ops);
+    TRACE_PRINT(retry);
+    TRACE_PRINT(contains);
+    TRACE_PRINT(traversal);
+    TRACE_PRINT(fail);
+    TRACE_PRINT(del)
+    TRACE_PRINT(ins);
+    TRACE_PRINT(load);
+    TRACE_PRINT(store);
+#undef TRACE_PRINT
+#define TRACE_PRINT(val) printf("%-10s: %ld\n", #val, val);
+    TRACE_PRINT(deletes);
+    TRACE_PRINT(inserts);
+#undef TRACE_PRINT
+    printf(
```

Comment:
Let's move the content for `printf`s to `README.md` file. So that, we can make program output more elegant.